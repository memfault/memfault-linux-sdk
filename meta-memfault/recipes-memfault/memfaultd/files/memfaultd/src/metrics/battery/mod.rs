//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod battery_reading_handler;
pub use battery_reading_handler::BatteryReadingHandler;

mod battery_monitor;
pub use battery_monitor::BatteryMonitor;
pub use battery_monitor::BatteryMonitorReading;
