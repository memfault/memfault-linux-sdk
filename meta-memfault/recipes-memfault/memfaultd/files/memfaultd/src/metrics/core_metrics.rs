//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::collections::HashSet;

use super::MetricStringKey;
use crate::util::wildcard_pattern::WildcardPattern;

// Connectivity metrics
pub const METRIC_MF_SYNC_SUCCESS: &str = "sync_memfault_successful";
pub const METRIC_MF_SYNC_FAILURE: &str = "sync_memfault_failure";
pub const METRIC_CONNECTED_TIME: &str = "connectivity_connected_time_ms";
pub const METRIC_EXPECTED_CONNECTED_TIME: &str = "connectivity_expected_time_ms";
pub const METRIC_SYNC_SUCCESS: &str = "sync_successful";
pub const METRIC_SYNC_FAILURE: &str = "sync_failure";
pub const METRIC_CONNECTIVITY_RECV_BYTES: &str = "connectivity_recv_bytes";
pub const METRIC_CONNECTIVITY_SENT_BYTES: &str = "connectivity_sent_bytes";
pub const METRIC_CONNECTIVITY_INTERFACE_RECV_BYTES_PREFIX: &str = "connectivity_";
pub const METRIC_CONNECTIVITY_INTERFACE_RECV_BYTES_SUFFIX: &str = "_recv_bytes";
pub const METRIC_CONNECTIVITY_INTERFACE_SENT_BYTES_PREFIX: &str = "connectivity_";
pub const METRIC_CONNECTIVITY_INTERFACE_SENT_BYTES_SUFFIX: &str = "_sent_bytes";

// Stability metrics
pub const METRIC_OPERATIONAL_HOURS: &str = "operational_hours";
pub const METRIC_OPERATIONAL_CRASHFREE_HOURS: &str = "operational_crashfree_hours";
pub const METRIC_OPERATIONAL_CRASHES: &str = "operational_crashes";
pub const METRIC_OPERATIONAL_CRASHES_PROCESS_PREFIX: &str = "operational_crashes_";

// Battery metrics
pub const METRIC_BATTERY_DISCHARGE_DURATION_MS: &str = "battery_discharge_duration_ms";
pub const METRIC_BATTERY_SOC_PCT_DROP: &str = "battery_soc_pct_drop";
pub const METRIC_BATTERY_SOC_PCT: &str = "battery_soc_pct";

// Memory Metrics
pub const METRIC_MEMORY_PCT: &str = "memory_pct";

// CPU Metrics
pub const METRIC_CPU_USAGE_PCT: &str = "cpu_usage_pct";

// Disk Space Metrics
pub const METRIC_STORAGE_USED_DISK_PCT_PREFIX: &str = "storage_used_";
pub const METRIC_STORAGE_USED_DISK_PCT_SUFFIX: &str = "_pct";

// Process Metrics
// Note: The metric keys for per-process core metrics follow the below
// format:
//     "<PREFIX><PROCESS NAME><SUFFIX>"
pub const METRIC_CPU_USAGE_PROCESS_PCT_PREFIX: &str = "cpu_usage_";
pub const METRIC_CPU_USAGE_PROCESS_PCT_SUFFIX: &str = "_pct";
pub const METRIC_MEMORY_PROCESS_PCT_PREFIX: &str = "memory_";
pub const METRIC_MEMORY_PROCESS_PCT_SUFFIX: &str = "_pct";

const SESSION_CORE_METRICS: &[&str; 12] = &[
    METRIC_MF_SYNC_FAILURE,
    METRIC_MF_SYNC_SUCCESS,
    METRIC_BATTERY_DISCHARGE_DURATION_MS,
    METRIC_BATTERY_SOC_PCT_DROP,
    METRIC_CONNECTED_TIME,
    METRIC_EXPECTED_CONNECTED_TIME,
    METRIC_SYNC_FAILURE,
    METRIC_SYNC_SUCCESS,
    METRIC_OPERATIONAL_CRASHES,
    METRIC_MEMORY_PCT,
    METRIC_CONNECTIVITY_RECV_BYTES,
    METRIC_CONNECTIVITY_SENT_BYTES,
];

const SESSION_CORE_WILDCARD_METIRCS: &[(&str, &str); 6] = &[
    (
        METRIC_CPU_USAGE_PROCESS_PCT_PREFIX,
        METRIC_CPU_USAGE_PROCESS_PCT_SUFFIX,
    ),
    (
        METRIC_MEMORY_PROCESS_PCT_PREFIX,
        METRIC_MEMORY_PROCESS_PCT_SUFFIX,
    ),
    (
        METRIC_STORAGE_USED_DISK_PCT_PREFIX,
        METRIC_STORAGE_USED_DISK_PCT_SUFFIX,
    ),
    (
        METRIC_CONNECTIVITY_INTERFACE_RECV_BYTES_PREFIX,
        METRIC_CONNECTIVITY_INTERFACE_RECV_BYTES_SUFFIX,
    ),
    (
        METRIC_CONNECTIVITY_INTERFACE_SENT_BYTES_PREFIX,
        METRIC_CONNECTIVITY_INTERFACE_SENT_BYTES_SUFFIX,
    ),
    (METRIC_OPERATIONAL_CRASHES_PROCESS_PREFIX, ""),
];

#[derive(Clone)]
pub struct CoreMetricKeys {
    pub string_keys: HashSet<MetricStringKey>,
    pub wildcard_pattern_keys: Vec<WildcardPattern>,
}

impl CoreMetricKeys {
    pub fn get_session_core_metrics() -> Self {
        let string_keys = HashSet::from_iter(SESSION_CORE_METRICS.map(MetricStringKey::from));
        let wildcard_pattern_keys = SESSION_CORE_WILDCARD_METIRCS
            .map(|(prefix, suffix)| WildcardPattern::new(prefix, suffix))
            .to_vec();

        Self {
            string_keys,
            wildcard_pattern_keys,
        }
    }
}
