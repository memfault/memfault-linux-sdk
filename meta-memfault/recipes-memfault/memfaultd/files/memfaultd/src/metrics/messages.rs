//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::path::PathBuf;

use eyre::Result;
use ssf::{Message, MsgMailbox};

use crate::network::NetworkConfig;

use super::{KeyedMetricReading, SessionName};

/// Allows KeyedMetricReading to be sent as a message. The `ssf` framework will
/// automatically support sending `Vec<KeyedMetricReading>` as well.
impl Message for KeyedMetricReading {
    type Reply = Result<()>;
}

/// Syntactic Sugar because this is the most-used type of mailbox in memfaultd.
pub type MetricsMBox = MsgMailbox<Vec<KeyedMetricReading>>;

pub enum SessionEventMessage {
    StartSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
    },
    StopSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
        mar_staging_area: PathBuf,
        network_config: NetworkConfig,
    },
}

impl Message for SessionEventMessage {
    type Reply = Result<()>;
}
