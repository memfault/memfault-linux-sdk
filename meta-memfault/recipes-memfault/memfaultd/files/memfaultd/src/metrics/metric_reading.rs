//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::str::FromStr;

use chrono::{Duration, Utc};
use eyre::{eyre, ErrReport};
use serde::{Deserialize, Serialize};

use super::{MetricStringKey, MetricTimestamp};

use crate::util::serialization::float_to_duration;

#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
pub enum MetricReading {
    /// TimeWeightedAverage readings will be aggregated based on
    /// the time the reading was captured over.
    TimeWeightedAverage {
        value: f64,
        timestamp: MetricTimestamp,
        /// Time period considered for this reading. This is only used to give a "time-weight" to the first
        /// value in the series (when using time-weighted averages). For future values we will use the time
        /// difference we measure between the two points         
        /// In doubt, it's safe to use Duration::from_secs(0) here. This means the first value will be ignored.
        #[serde(with = "float_to_duration")]
        interval: Duration,
    },
    /// A non-decreasing monotonic sum. Within a metric report, Counter readings are summed together.
    Counter {
        value: f64,
        timestamp: MetricTimestamp,
    },
    /// Gauges are absolute values. We keep the latest value collected during a metric report.
    Gauge {
        value: f64,
        timestamp: MetricTimestamp,
    },
    /// Histogram readings are averaged together by dividing the sum of the values
    /// by the number of readings over the duration of a metric report.
    Histogram {
        value: f64,
        timestamp: MetricTimestamp,
    },
}

#[derive(Debug, Serialize, Clone, Deserialize)]
pub struct KeyedMetricReading {
    pub name: MetricStringKey,
    pub value: MetricReading,
}

impl KeyedMetricReading {
    pub fn new(name: MetricStringKey, value: MetricReading) -> Self {
        Self { name, value }
    }
}

impl FromStr for KeyedMetricReading {
    type Err = ErrReport;

    /// Deserialize a string in the form <MetricStringKey>=<f64> to a Gauge metric reading
    ///
    /// Currently deserialization to a KeyedMetricReading with any other type of MetricReading
    /// as its value is not supported
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (key, value_str) = s.split_once('=').ok_or(eyre!(
            "Gauge metric reading should be specified as KEY=VALUE"
        ))?;

        // Let's ensure the key is valid first:
        let metric_key = MetricStringKey::from_str(key).map_err(|e| eyre!(e))?;
        let value =
            f64::from_str(value_str).map_err(|e| eyre!("Invalid value {}: {}", value_str, e))?;

        let reading = MetricReading::Gauge {
            value,
            timestamp: Utc::now(),
        };
        Ok(KeyedMetricReading::new(metric_key, reading))
    }
}
