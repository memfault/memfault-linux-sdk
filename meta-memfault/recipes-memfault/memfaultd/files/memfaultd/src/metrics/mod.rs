//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::{DateTime, Utc};

mod battery;
pub use battery::BatteryMonitor;
pub use battery::BatteryMonitorReading;
pub use battery::BatteryReadingHandler;

mod connectivity;
pub use connectivity::ConnectivityMonitor;
pub use connectivity::ReportSyncEventHandler;

mod metric_string_key;
pub use metric_string_key::MetricStringKey;

mod metric_report;
pub use metric_report::MetricReport;
pub use metric_report::MetricReportType;

mod metric_report_manager;
pub use metric_report_manager::MetricReportManager;
pub use metric_report_manager::TakeMetrics;

mod messages;
pub use messages::*;

mod metric_reading;
pub use metric_reading::KeyedMetricReading;
pub use metric_reading::MetricReading;

mod timeseries;

mod metric_value;
pub use metric_value::MetricValue;

pub type MetricTimestamp = DateTime<Utc>;

mod periodic_metric_report;
pub use periodic_metric_report::PeriodicMetricReportDumper;

mod crashfree_interval;
pub use crashfree_interval::CrashFreeIntervalTracker;

mod session_name;
pub use session_name::SessionName;

mod session_event_handler;
pub use session_event_handler::SessionEventHandler;

pub mod core_metrics;

pub mod statsd_server;
pub use statsd_server::StatsDServer;

mod system_metrics;
pub use system_metrics::SystemMetricsCollector;
pub use system_metrics::BUILTIN_SYSTEM_METRIC_NAMESPACES;
