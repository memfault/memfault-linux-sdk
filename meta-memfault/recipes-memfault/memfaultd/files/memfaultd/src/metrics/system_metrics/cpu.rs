//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Collect CPU metric readings from /proc/stat
//!
//! This module parses CPU statistics from /proc/stat and
//! constructs KeyedMetricReadings based on those statistics.
//! Because the /proc/stat values are accumulations since boot,
//! a "previous reading" (stored in CpuMetricCollector) is
//! required to calculate the time each CPU core
//! has spent in each state since the last reading.
//!
//! Example /proc/stat contents:
//! cpu  326218 0 178980 36612114 6054 0 11961 0 0 0
//! cpu0 77186 0 73689 9126238 1353 0 6352 0 0 0
//! cpu1 83902 0 35260 9161039 1524 0 1865 0 0 0
//! cpu2 83599 0 35323 9161010 1676 0 1875 0 0 0
//! cpu3 81530 0 34707 9163825 1500 0 1867 0 0 0
//! intr 95400676 0 9795 1436573 0 0 0 0 0 0 0 0 93204555 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 77883 0 530 0 0 1523 0 0 468762 0 0 97412 103573 0 70 0 0 0 0 0
//! ctxt 9591503
//! btime 1714309294
//! processes 9416
//! procs_running 1
//! procs_blocked 0
//! softirq 47765068 15 3173702 0 541726 82192 0 1979 41497887 0 2467567
//!
//! Only the lines that start with "cpu" are currently
//! processed into metric readings by this module - the rest are discarded.
//!
//! See additional Linux kernel documentation on /proc/stat here:
//! https://docs.kernel.org/filesystems/proc.html#miscellaneous-kernel-statistics-in-proc-stat
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::iter::zip;
use std::ops::Add;
use std::path::Path;
use std::str::FromStr;

use chrono::Utc;
use nom::{
    bytes::complete::tag,
    character::complete::{digit0, space1},
    multi::count,
    number::complete::double,
    sequence::{pair, preceded},
    IResult,
};

use crate::metrics::{KeyedMetricReading, MetricReading, MetricStringKey};
use eyre::{eyre, ErrReport, Result};

const PROC_STAT_PATH: &str = "/proc/stat";
pub const CPU_METRIC_NAMESPACE: &str = "cpu";

pub struct CpuMetricCollector {
    last_reading_by_cpu: HashMap<String, Vec<f64>>,
}

impl CpuMetricCollector {
    pub fn new() -> Self {
        Self {
            last_reading_by_cpu: HashMap::new(),
        }
    }

    pub fn get_cpu_metrics(&mut self) -> Result<Vec<KeyedMetricReading>> {
        // Track if any lines in /proc/stat are parse-able
        // so we can alert user if none are
        let mut no_parseable_lines = true;

        let path = Path::new(PROC_STAT_PATH);

        let file = File::open(path)?;
        let reader = BufReader::new(file);

        let mut cpu_metric_readings = vec![];

        for line in reader.lines() {
            // Discard errors - the assumption here is that we are only parsing
            // lines that follow the specified format and expect other lines in the file to error
            if let Ok((cpu_id, cpu_stats)) = Self::parse_proc_stat_line_cpu(line?.trim()) {
                no_parseable_lines = false;
                if let Ok(Some(mut readings)) = self.delta_since_last_reading(cpu_id, cpu_stats) {
                    cpu_metric_readings.append(&mut readings);
                }
            }
        }

        // Check if we were able to parse at least one CPU metric reading
        if !no_parseable_lines {
            Ok(cpu_metric_readings)
        } else {
            Err(eyre!(
                    "No CPU metrics were collected from {} - is it a properly formatted /proc/stat file?",
                    PROC_STAT_PATH
            ))
        }
    }

    /// Parse a cpu ID from a line of /proc/stat
    ///
    /// A cpu ID is the digit following the 3 character string "cpu"
    /// No ID being present implies these stats are for the total
    /// of all cores on the CPU
    fn parse_cpu_id(input: &str) -> IResult<&str, &str> {
        preceded(tag("cpu"), digit0)(input)
    }

    /// Parse the CPU stats from the suffix of a /proc/stat line following the cpu ID
    ///
    /// 7 or more space delimited integers are expected. Values after the 7th are discarded.
    fn parse_cpu_stats(input: &str) -> IResult<&str, Vec<f64>> {
        count(preceded(space1, double), 7)(input)
    }

    /// Parse the output of a line of /proc/stat, returning
    /// a pair of the cpu ID that the parsed line corresponds
    /// to and the first 7 floats listed for it
    ///
    /// The 7 floats represent how much time since boot the cpu has
    /// spent in the "user", "nice", "system", "idle", "iowait", "irq",
    /// "softirq", in that order    
    /// Example of a valid parse-able line:
    /// cpu2 36675 176 11216 1552961 689 0 54
    fn parse_proc_stat_line_cpu(line: &str) -> Result<(String, Vec<f64>)> {
        let (_remaining, (cpu_id, cpu_stats)) =
            pair(Self::parse_cpu_id, Self::parse_cpu_stats)(line)
                .map_err(|_e| eyre!("Failed to parse CPU stats line: {}", line))?;
        Ok(("cpu".to_string().add(cpu_id), cpu_stats))
    }

    /// Calculate the time spent in each state for the
    /// provided CPU core since the last reading collected
    /// by the CpuMetricCollector
    ///
    /// Returns an Ok(None) if there is no prior reading
    /// to calculate a delta from.
    fn delta_since_last_reading(
        &mut self,
        cpu_id: String,
        cpu_stats: Vec<f64>,
    ) -> Result<Option<Vec<KeyedMetricReading>>> {
        // Check to make sure there was a previous reading to calculate a delta with
        if let Some(last_stats) = self
            .last_reading_by_cpu
            .insert(cpu_id.clone(), cpu_stats.clone())
        {
            let delta = cpu_stats
                .iter()
                .zip(last_stats)
                .map(|(current, previous)| current - previous);

            let cpu_states_with_ticks = zip(
                ["user", "nice", "system", "idle", "iowait", "irq", "softirq"],
                delta,
            )
            .collect::<Vec<(&str, f64)>>();

            let sum: f64 = cpu_states_with_ticks.iter().map(|(_k, v)| v).sum();
            let timestamp = Utc::now();

            let readings = cpu_states_with_ticks
                .iter()
                .map(|(key, value)| -> Result<KeyedMetricReading, ErrReport> {
                    Ok(KeyedMetricReading::new(
                        MetricStringKey::from_str(&format!(
                            "{}/{}/percent/{}",
                            CPU_METRIC_NAMESPACE, cpu_id, key
                        ))
                        .map_err(|e| eyre!(e))?,
                        MetricReading::Histogram {
                            // Transform raw tick value to a percentage
                            value: 100.0 * value / sum,
                            timestamp,
                        },
                    ))
                })
                .collect::<Result<Vec<KeyedMetricReading>>>();
            match readings {
                Ok(readings) => Ok(Some(readings)),
                Err(e) => Err(e),
            }
        } else {
            Ok(None)
        }
    }
}

#[cfg(test)]
mod test {

    use insta::{assert_json_snapshot, rounded_redaction, with_settings};
    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case("cpu 1000 5 0 0 2 0 0", "test_basic_line")]
    #[case("cpu0 1000 5 0 0 2 0 0 0 0 0", "test_basic_line_with_extra")]
    fn test_process_valid_proc_stat_line(#[case] proc_stat_line: &str, #[case] test_name: &str) {
        assert_json_snapshot!(test_name, 
                              CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line).unwrap(), 
                              {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
    }

    #[rstest]
    #[case("cpu 1000 5 0 0 2")]
    #[case("1000 5 0 0 2 0 0 0 0 0")]
    #[case("processor0 1000 5 0 0 2 0 0 0 0 0")]
    #[case("softirq 403453672 10204651 21667771 199 12328940 529390 0 3519783 161759969 147995 193294974")]
    fn test_fails_on_invalid_proc_stat_line(#[case] proc_stat_line: &str) {
        assert!(CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line).is_err())
    }

    #[rstest]
    #[case(
        "cpu 1000 5 0 0 2 0 0",
        "cpu 1500 20 4 1 2 0 0",
        "cpu 1550 200 40 3 3 0 0",
        "basic_delta"
    )]
    fn test_cpu_metric_collector_calcs(
        #[case] proc_stat_line_a: &str,
        #[case] proc_stat_line_b: &str,
        #[case] proc_stat_line_c: &str,
        #[case] test_name: &str,
    ) {
        let mut cpu_metric_collector = CpuMetricCollector::new();

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_a).unwrap();
        let result_a = cpu_metric_collector.delta_since_last_reading(cpu, stats);
        matches!(result_a, Ok(None));

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_b).unwrap();
        let result_b = cpu_metric_collector.delta_since_last_reading(cpu, stats);

        assert!(result_b.is_ok());

        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(format!("{}_{}", test_name, "a_b_metrics"),
                                  result_b.unwrap(),
                                  {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
        });

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_c).unwrap();
        let result_c = cpu_metric_collector.delta_since_last_reading(cpu, stats);

        assert!(result_c.is_ok());

        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(format!("{}_{}", test_name, "b_c_metrics"),
                                  result_c.unwrap(),
                                  {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
        });
    }

    #[rstest]
    #[case(
        "cpu1 40 20 30 10 0 0 0",
        "cpu0 1500 20 4 1 2 0 0",
        "cpu1 110 30 40 12 5 3 0",
        "different_cores"
    )]
    fn test_cpu_metric_collector_different_cores(
        #[case] proc_stat_line_a: &str,
        #[case] proc_stat_line_b: &str,
        #[case] proc_stat_line_c: &str,
        #[case] test_name: &str,
    ) {
        let mut cpu_metric_collector = CpuMetricCollector::new();

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_a).unwrap();
        let result_a = cpu_metric_collector.delta_since_last_reading(cpu, stats);
        matches!(result_a, Ok(None));

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_b).unwrap();
        let result_b = cpu_metric_collector.delta_since_last_reading(cpu, stats);
        matches!(result_b, Ok(None));

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_c).unwrap();
        let result_c = cpu_metric_collector.delta_since_last_reading(cpu, stats);

        assert!(result_c.is_ok());

        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(format!("{}_{}", test_name, "a_c_metrics"),
                                  result_c.unwrap(),
                                  {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
        });
    }
}
