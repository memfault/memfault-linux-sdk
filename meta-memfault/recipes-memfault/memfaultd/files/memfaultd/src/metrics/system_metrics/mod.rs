//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    thread::sleep,
    time::{Duration, Instant},
};

use eyre::Result;
use log::{debug, warn};

use crate::{
    config::SystemMetricConfig,
    metrics::KeyedMetricReading,
    util::system::{bytes_per_page, clock_ticks_per_second},
};

mod cpu;
use crate::metrics::system_metrics::cpu::{CpuMetricCollector, CPU_METRIC_NAMESPACE};

mod thermal;
use crate::metrics::system_metrics::thermal::{ThermalMetricsCollector, THERMAL_METRIC_NAMESPACE};

mod memory;
use crate::metrics::system_metrics::memory::{MemoryMetricsCollector, MEMORY_METRIC_NAMESPACE};

mod network_interfaces;
use crate::metrics::system_metrics::network_interfaces::{
    NetworkInterfaceMetricCollector, NetworkInterfaceMetricsConfig,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
};

mod processes;
use processes::{ProcessMetricsCollector, PROCESSES_METRIC_NAMESPACE};

mod disk_space;
use disk_space::{
    DiskSpaceMetricCollector, DiskSpaceMetricsConfig, NixStatvfs, DISKSPACE_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
};

use self::processes::ProcessMetricsConfig;
use super::MetricsMBox;

pub const BUILTIN_SYSTEM_METRIC_NAMESPACES: &[&str; 7] = &[
    CPU_METRIC_NAMESPACE,
    MEMORY_METRIC_NAMESPACE,
    THERMAL_METRIC_NAMESPACE,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
    PROCESSES_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE,
    // Include in list of namespaces so that
    // legacy collectd from the "df" plugin
    // are still filtered out
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
];

pub trait SystemMetricFamilyCollector {
    fn collect_metrics(&mut self) -> Result<Vec<KeyedMetricReading>>;
    fn family_name(&self) -> &'static str;
}

pub struct SystemMetricsCollector {
    metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>>,
    metrics_mbox: MetricsMBox,
}

impl SystemMetricsCollector {
    pub fn new(system_metric_config: SystemMetricConfig, metrics_mbox: MetricsMBox) -> Self {
        // CPU, Memory, and Thermal metrics are captured by default
        let mut metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>> = vec![
            Box::new(CpuMetricCollector::new()),
            Box::new(MemoryMetricsCollector::new()),
            Box::new(ThermalMetricsCollector::new()),
        ];

        // Check if process metrics have been manually configured
        match system_metric_config.processes {
            Some(processes) if !processes.is_empty() => {
                metric_family_collectors.push(Box::new(ProcessMetricsCollector::<Instant>::new(
                    ProcessMetricsConfig::Processes(processes),
                    clock_ticks_per_second() as f64 / 1000.0,
                    bytes_per_page() as f64,
                )))
            }
            // Monitoring no processes means this collector is disabled
            Some(_empty_set) => {}
            None => {
                metric_family_collectors.push(Box::new(ProcessMetricsCollector::<Instant>::new(
                    ProcessMetricsConfig::Auto,
                    clock_ticks_per_second() as f64 / 1000.0,
                    bytes_per_page() as f64,
                )))
            }
        };

        // Check if disk space metrics have been manually configured
        match system_metric_config.disk_space {
            Some(disks) if !disks.is_empty() => {
                metric_family_collectors.push(Box::new(DiskSpaceMetricCollector::new(
                    NixStatvfs::new(),
                    DiskSpaceMetricsConfig::Disks(disks),
                )))
            }
            // Monitoring no disks means this collector is disabled
            Some(_empty_set) => {}
            None => metric_family_collectors.push(Box::new(DiskSpaceMetricCollector::new(
                NixStatvfs::new(),
                DiskSpaceMetricsConfig::Auto,
            ))),
        };

        // Check if network interface metrics have been manually configured
        match system_metric_config.network_interfaces {
            Some(interfaces) if !interfaces.is_empty() => metric_family_collectors.push(Box::new(
                NetworkInterfaceMetricCollector::<Instant>::new(
                    NetworkInterfaceMetricsConfig::Interfaces(interfaces),
                ),
            )),
            // Monitoring no interfaces means this collector is disabled
            Some(_empty_set) => {}
            None => metric_family_collectors.push(Box::new(NetworkInterfaceMetricCollector::<
                Instant,
            >::new(
                NetworkInterfaceMetricsConfig::Auto
            ))),
        };

        Self {
            metric_family_collectors,
            metrics_mbox,
        }
    }

    pub fn run(&mut self, metric_poll_duration: Duration) {
        loop {
            for collector in self.metric_family_collectors.iter_mut() {
                match collector.collect_metrics() {
                    Ok(readings) => {
                        if let Err(e) = self.metrics_mbox.send_and_forget(readings) {
                            debug!(
                                "Couldn't add metric reading for family \"{}\": {:?}",
                                collector.family_name(),
                                e
                            )
                        }
                    }
                    Err(e) => warn!(
                        "Failed to collect readings for family \"{}\": {}",
                        collector.family_name(),
                        e
                    ),
                }
            }

            sleep(metric_poll_duration);
        }
    }
}
