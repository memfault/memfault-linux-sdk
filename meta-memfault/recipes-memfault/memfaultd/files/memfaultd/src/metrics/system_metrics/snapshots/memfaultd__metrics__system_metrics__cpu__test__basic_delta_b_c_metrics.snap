---
source: memfaultd/src/metrics/system_metrics/cpu.rs
expression: result_c
---
[
  {
    "name": "cpu/cpu/percent/idle",
    "value": {
      "Histogram": {
        "value": 0.74349,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "cpu/cpu/percent/iowait",
    "value": {
      "Histogram": {
        "value": 0.37175,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "cpu/cpu/percent/irq",
    "value": {
      "Histogram": {
        "value": 0.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "cpu/cpu/percent/nice",
    "value": {
      "Histogram": {
        "value": 66.9145,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "cpu/cpu/percent/softirq",
    "value": {
      "Histogram": {
        "value": 0.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "cpu/cpu/percent/system",
    "value": {
      "Histogram": {
        "value": 13.3829,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "cpu/cpu/percent/user",
    "value": {
      "Histogram": {
        "value": 18.58736,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "cpu_usage_pct",
    "value": {
      "Histogram": {
        "value": 99.25651,
        "timestamp": "[timestamp]"
      }
    }
  }
]
